### PR TITLE
specs: add QGB specs

### PR DESCRIPTION
**Pull request by SweeXordious**
## Overview

Closes https://github.com/celestiaorg/celestia-app/issues/1872

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

### CODE REVIEW

File: x/qgb/README.md
Comment (cmwaters) [inline x/qgb/README.md:131]: Why is the square size important?

File: x/qgb/README.md
Comment (cmwaters) [inline x/qgb/README.md:407]: Do we have links for these?

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:7]: > Does the orchestrator actually sign the data commitments or the validator?

The validator runs the orchestrator software

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:17]: but this part is only for definitions. Do you mean we should add a link for that section in the md file?

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:19]: It could, but we don't do that at the moment. Every nonce is processed separately without any order.
We could sign nonce 10 then 2 then 11...

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:197]: We could do that, but we don't want to expect from a validator to sign if he left the validator set. Also, when we will want to implement slashing, it wouldn't make sense to slash for liveness for a validator that started  unbonding

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:203]: No upper bound. And no rationale, inherited from the gravity bridge

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:407]: good catch :+1:

File: x/qgb/README.md
Comment (cmwaters) [reply x/qgb/README.md:197]: Even with this logic, we'd still expect situations where we're asking a validator to sign an attestation even although they are not part of the validator set. Imagine if they didn't unbond or anything but we're simply kicked out because another validator delegated enough to get itself in the set and kick out the lowest staked validator

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:131]: I didn't want to add it because there isn't any write-up for it so far.  I will open an issue to add it

Comment (SweeXordious) [reply x/qgb/README.md:131]: https://github.com/celestiaorg/celestia-app/issues/1918

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:197]: Got it, but let's imagine the case where one validator which has 15% of the staking power leaves the validator set, then another validator joins with 15%. Then, if we don't have the unbonding hook, no change will happen to the valset because there was no power change

File: x/qgb/README.md
Comment (cmwaters) [reply x/qgb/README.md:7]: Ok, that's a bit confusing

File: x/qgb/README.md
Comment (cmwaters) [reply x/qgb/README.md:197]: Ah measurement of the power change is net? That doesn't sound too safe/correct

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:7]: We still can rename tho

File: x/qgb/README.md
Comment (cmwaters) [reply x/qgb/README.md:203]: I guess 5% errs on the side of caution but the upper bound might be something like 33% or 1/3. I haven't spent too much time thinking about the attack vectors here but it might be good to eventually write them up as a form of rationale

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:197]: It wouldn't need to be caught because its just a redelegation. 
We're even normalizing the powers to avoid creating valsets in the case of inflation or...

File: x/qgb/README.md
Comment (cmwaters) [inline x/qgb/README.md:407]:
```suggestion
For QGB v2, the orchestrator and relayer implementations are in the [orchestratory-relayer](https://github.com/celestiaorg/orchestrator-relayer) repo.
```

File: x/qgb/README.md
Comment (cmwaters) [inline x/qgb/README.md:229]: Is there a threshold for how much they should unbond before this gets triggered?

**Review by cmwaters — DISMISSED** [review]
Documentation looks overall clear and helpful. I still have a few questions about the protocol itself but perhaps that can be addressed when we start the audit

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:229]: There is no threshold, the time a validator starts unbonding, this gets triggered

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:203]: I will open an issue to think on an upper bound for this value, thanks for pointing this out

Comment (SweeXordious) [reply x/qgb/README.md:203]: https://github.com/celestiaorg/celestia-app/issues/1920

File: x/qgb/README.md
Comment (evan-forbes) [inline x/qgb/README.md:285]: [discusison around planning]
certainly not something we do in this PR, but as we've discussed in the past, we will eventually need to nail down the UX. All the tool that are needed are in these commands, but devs don't really have easy access to them. 

If they're looking at these specs, I would feel a bit confused if this were the only api, since these just logs a yes or no.

File: x/qgb/README.md
Comment (evan-forbes) [inline x/qgb/README.md:267]: [not blocking suggestion]

imo we should be diligent in removing things that do not directly contribute to this module, we might even want to remove the other flags/aliases section.
```suggestion
```

File: x/qgb/README.md
Comment (evan-forbes) [inline x/qgb/README.md:244]: >  INT ✘ 

what's this?

Comment (evan-forbes) [inline x/qgb/README.md:244]: as also discussed with the comment on flags, I think it would be cleaner to just list the commands and a short description perhaps, and users can call `--help` if they want to view other flags. This also allows us to avoid manually updating this when those things change

File: x/qgb/README.md
Comment (evan-forbes) [inline x/qgb/README.md:409]: do we provide other information on versions elsewhere? afaik, I think we're the only ones who know what v1 or v2 are :sweat_smile: . If we don't distinguish between v1 and v2 in the code, we might want to only reference off hand here.

if we want to document the iterations we've done, then we might want to just do that in a short paragraph here, or in a separate document with more details. or maybe reference old ADRS or something not sure

**Review by evan-forbes — COMMENTED** [review]
nice update! I think this will pair well with a higher level (but still very technical) document about trust assumptions, broader scope/connection contracts, etc whenever we get around to it.

no blocking comments, just some questions. I'll take a second review after chewing on it as well

File: x/qgb/README.md
Comment (cmwaters) [reply x/qgb/README.md:197]: Ok so just to confirm power change isn't just the total power now compared to the total power before? It actually looks to see what validators had what power?

File: x/qgb/README.md
Comment (cmwaters) [reply x/qgb/README.md:229]: So if a validator unbonds a single TIA token this will trigger? (Even though there would be no noticeable change to the validator set)

Comment (SweeXordious) [reply x/qgb/README.md:229]: In this case yes, because now he has lower power in the network, and it might make a difference when calculating the 2/3 relaying threshold

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:63]: The percentage change reduces the frequency in case of redelegations for example.
The normalization reduces the frequency in case of, for example, inflation:
```
if there is a small increase in the total on-chain voting power due to inflation, there is no need to create a new validator set. This is because the relative proportions of the validators remain the same, and the normalized QGB power doesn't show any significant difference.
```

File: x/qgb/README.md
Comment (SweeXordious) [inline x/qgb/README.md:73]:
```suggestion
A data commitment is an attestation type representing a request to commit over a set of blocks. It provides an end exclusive range of blocks for orchestrators to sign over and propagate in the QGB P2P network. The range is defined by the param [`DataCommitmentWindow`](https://github.com/celestiaorg/celestia-app/blob/fc83b04c3a5638ac8d415770e38a4046b84fa128/x/qgb/keeper/keeper_data_commitment.go#L44-L50), more on this below.
```

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md]: good catch [c7e4a14](https://github.com/celestiaorg/celestia-app/pull/1911/commits/c7e4a144724941f118b321fce9053ab11bc17aa7)

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:77]: https://github.com/celestiaorg/celestia-app/pull/1911#discussion_r1231098193

File: x/qgb/README.md
Comment (cmwaters) [reply x/qgb/README.md:197]: So is the unbonding hook still necessary?

File: x/qgb/README.md
Comment (cmwaters) [reply x/qgb/README.md:229]: But why is a validator unbonding a single TIA token different to a delegator unbonding a single TIA token from the same validator. They both have the same voting power difference. I don't see we can't just use the significant voting power threshold as the sole trigger for a valset attestation.

Couldn't this also be some form of spam. i.e. every height I unbond 1 TIA, meaning that an attestation is made every height.

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:197]: yes, because if that sum is not significant, no change to the valset will happen, even if the validator unbonded. Meaning we will still keep waiting for them to sign or think we can slash them

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:15]: yes

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:229]: > I don't see we can't just use the significant voting power threshold as the sole trigger for a valset attestation.

We could but we won't be catching the case where a validator unbonds all his tokens, which are `4.9999999` of the power, and he would be able to double sign afterward without us slashing him.

> Couldn't this also be some form of spam. i.e. every height I unbond 1 TIA, meaning that an attestation is made every height.

They would also be paying the transaction fees. And we have pruning, so it would be around 500mb if they create a valset every block for 3 weeks for 15sec block time. 500mb of state that is not really manipulated inside the state machine that much is okey, I guess

Comment (cmwaters) [reply x/qgb/README.md:229]: > We could but we won't be catching the case where a validator unbonds all his tokens, which are 4.9999999 of the power, and he would be able to double sign afterward without us slashing him.

Don't you require 1/3 to double sign in order to genuinely equivocate (I'm just assuming light client verification security. I actually don't know the specifics)

Also it will take 21 days for the validator to truly unbond (or is the unbond hook for when their tokens are unlocked).

Comment (SweeXordious) [reply x/qgb/README.md:229]: > Don't you require 1/3 to double sign in order to genuinely equivocate

Yes, but even if they will be unable to halt the QGB, we still would want to slash any bad behavior

> Also it will take 21 days for the validator to truly unbond (or is the unbond hook for when their tokens are unlocked).

But, if after they start unbonding their tokens, we don't create a valset. Let's imagine a scenario where during that 21 days, no change happens to the valset. Then, day 22 we will still have the relayer getting the signature of that 
 malicious orchestrator, his tokens will be unlocked, and he will be able to submit any signature and will get relayed to the QGB. Of course, we're implementing the relayer to be smarter and not relay it, but also we wouldn't want this behaviour to go unpunished

File: x/qgb/README.md
Comment (evan-forbes) [reply x/qgb/README.md:63]: cool that part makes sense, and including the rational is a good thing.  The part I think we need to change is to also discuss the percentage change, which is the driver of the reduction. We only change if there is a 5% difference. We calculate that difference by first normalizing. This combined mechanism is responsible for safely reducing the frequency of valdiator set updates.

File: x/qgb/README.md
Comment (evan-forbes) [reply x/qgb/README.md:387]: sounds good

File: x/qgb/README.md
Comment (evan-forbes) [reply x/qgb/README.md:77]: cool

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md]: [cca1dab](https://github.com/celestiaorg/celestia-app/pull/1911/commits/cca1dab597b78af7c4e6d38a624e906301e7a141)

File: x/qgb/README.md
Comment (cmwaters) [reply x/qgb/README.md:231]: > Yes, but even if they will be unable to halt the QGB, we still would want to slash any bad behavior

I'm usually a punishment minimalist. So long as there's not a 1/3 validator set change, there's nothing a malicious validator can do. I think 5% is a good threshold in this case.

**Review by cmwaters — APPROVED** [review]
Thanks for answering all my questions. 👍

File: x/qgb/README.md
Comment (evan-forbes) [inline x/qgb/README.md:379]: adding this, but tbh probably not worth the extra round of reviews lol

used 3 " " instead of 4 since the rest are using 3
```suggestion
   panic("hooks initialized before GravityKeeper")
```

**Review by evan-forbes — APPROVED** [review]
dope I think this pairs really well with #1974!

File: x/qgb/README.md
Comment (SweeXordious) [reply x/qgb/README.md:379]: I will handle it in a subsequent PR, thanks :+1:

Comment (SweeXordious) [reply x/qgb/README.md:379]: https://github.com/celestiaorg/celestia-app/pull/1976

**Pull request merged**
